agent/auth/auth.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Ultraviolet
+// SPDX-License-Identifier: Apache-2.0
+
+package auth
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+
+	"github.com/ultravioletrs/cocos/agent"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+type UserRole string
+
+const (
+	UserMetadataKey                = "user-id"
+	SignatureMetadataKey           = "signature"
+	consumerRole          UserRole = "consumer"
+	dataProviderRole      UserRole = "data-provider"
+	algorithmProviderRole UserRole = "algorithm-provider"
+)
+
+var errNotRSAPublicKey = errors.New("not an RSA public key")
+
+type Service struct {
+	users map[string]struct {
+		PublicKey *rsa.PublicKey
+		Role      UserRole
+	}
+}
+
+func New(manifest agent.Computation) (*Service, error) {
+	s := &Service{
+		users: make(map[string]struct {
+			PublicKey *rsa.PublicKey
+			Role      UserRole
+		}),
+	}
+	for _, rc := range manifest.ResultConsumers {
+		pubKey, err := x509.ParsePKIXPublicKey(rc.UserKey)
+		if err != nil {
+			return nil, err
+		}
+		rsaPubKey, ok := pubKey.(*rsa.PublicKey)
+		if !ok {
+			return nil, errNotRSAPublicKey
+		}
+		s.users[rc.Consumer] = struct {
+			PublicKey *rsa.PublicKey
+			Role      UserRole
+		}{
+			PublicKey: rsaPubKey,
+			Role:      consumerRole,
+		}
+	}
+	for _, dp := range manifest.Datasets {
+		pubKey, err := x509.ParsePKIXPublicKey(dp.UserKey)
+		if err != nil {
+			return nil, err
+		}
+		rsaPubKey, ok := pubKey.(*rsa.PublicKey)
+		if !ok {
+			return nil, errNotRSAPublicKey
+		}
+		s.users[dp.Provider] = struct {
+			PublicKey *rsa.PublicKey
+			Role      UserRole
+		}{
+			PublicKey: rsaPubKey,
+			Role:      dataProviderRole,
+		}
+	}
+	pubKey, err := x509.ParsePKIXPublicKey(manifest.Algorithm.UserKey)
+	if err != nil {
+		return nil, err
+	}
+	rsaPubKey, ok := pubKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, errNotRSAPublicKey
+	}
+	s.users[manifest.Algorithm.Provider] = struct {
+		PublicKey *rsa.PublicKey
+		Role      UserRole
+	}{
+		PublicKey: rsaPubKey,
+		Role:      algorithmProviderRole,
+	}
+	return s, nil
+}
+
+func (s *Service) AuthStreamInterceptor() grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		switch info.FullMethod {
+		case agent.AgentService_Algo_FullMethodName, agent.AgentService_Data_FullMethodName:
+		default:
+			return handler(srv, stream)
+		}
+		md, ok := metadata.FromIncomingContext(stream.Context())
+		if !ok {
+			return status.Errorf(codes.Unauthenticated, "missing metadata")
+		}
+		userID, signature, err := extractSignatureAndUserID(md)
+		if err != nil {
+			return status.Errorf(codes.Unauthenticated, "invalid metadata")
+		}
+
```

Comment:
If you added a white line here for readability, add it consistently to the upper ifs as well